cmd/check_spec.go

```diff
@@ -122,58 +92,448 @@ func newCheckSpec(ctx context.Context) (*checkSpec, error) {
+		if err != nil {
+			printError("%v: unable to get an account\n", err)
+			markAllValidationsFailed(output)
+			return output
+		}
+		if err != nil {
+			printError("%v\n", errAccountNullPointer)
+			markAllValidationsFailed(output)
+			return output
+		}
 
-	// TODO: more checks
+		block, cs, _, fetchErr := cs.onlineFetcher.AccountCoinsRetry(
+			ctx,
+			Config.Network,
+			acct,
+			false,
+			currencies)
+		if fetchErr != nil {
+			printError("%v: unable to get coins for account: %v\n", fetchErr.Err, *acct)
+			markAllValidationsFailed(output)
+			return output
+		}
+
+		// block_identifier is required
+		err = validateBlockIdentifier(block)
+		if err != nil {
+			printError("%v\n", err)
+			setValidationStatus(output, blockID, checkSpecFailure)
+		}
+
+		// coins is required
+		err = validateCoins(cs)
+		if err != nil {
+			printError("%v\n", err)
+			setValidationStatus(output, coins, checkSpecFailure)
+		}
+	}
+
+	return output
+}
+
+func (cs *checkSpec) block(ctx context.Context) checkSpecOutput {
+	printInfo("validating /block ...\n")
+	output := checkSpecOutput{
+		api: block,
+		validation: map[checkSpecRequirement]checkSpecStatus{
+			idempotent: checkSpecSuccess,
+			defaultTip: checkSpecSuccess,
+		},
+	}
+	defer printInfo("/block validated\n")
+
+	res, fetchErr := cs.onlineFetcher.NetworkStatusRetry(ctx, Config.Network, nil)
+	if fetchErr != nil {
+		printError("%v: unable to get network status\n", fetchErr.Err)
+		markAllValidationsFailed(output)
+		return output
+	}
+
+	// multiple calls with the same hash should return the same block
+	var block *types.Block
+	var tipDiff int64 = 3
+	tip := res.CurrentBlockIdentifier
+	callTimes := 3
+
+	for i := 0; i < callTimes; i++ {
+		blockID := types.PartialBlockIdentifier{
+			Hash: &tip.Hash,
+		}
+		b, fetchErr := cs.onlineFetcher.BlockRetry(ctx, Config.Network, &blockID)
+		if fetchErr != nil {
+			printError("%v: unable to fetch block %v\n", fetchErr.Err, blockID)
+			markAllValidationsFailed(output)
+			return output
+		}
+
+		if block == nil {
+			block = b
+		} else if !isEqual(types.Hash(*block), types.Hash(*b)) {
+			printError("%v\n", errBlockNotIdempotent)
+			setValidationStatus(output, idempotent, checkSpecFailure)
+		}
+	}
+
+	// tip shoud be returned if block_identifier is not specified
+	emptyBlockID := &types.PartialBlockIdentifier{}
+	block, fetchErr = cs.onlineFetcher.BlockRetry(ctx, Config.Network, emptyBlockID)
+	if fetchErr != nil {
+		printError("%v: unable to fetch tip block\n", fetchErr.Err)
+		setValidationStatus(output, defaultTip, checkSpecFailure)
+		return output
+	}
+
+	// fetch the tip block again
+	res, fetchErr = cs.onlineFetcher.NetworkStatusRetry(ctx, Config.Network, nil)
+	if fetchErr != nil {
+		printError("%v: unable to get network status\n", fetchErr.Err)
+		setValidationStatus(output, defaultTip, checkSpecFailure)
+		return output
+	}
+	tip = res.CurrentBlockIdentifier
+	if tip.Index-block.BlockIdentifier.Index > tipDiff {
```

Comment:
Can you explain why

Reply:
Yes. This piece of code is to verify that tip should be returned when /block request is empty, so we follow those steps:

- Get current tip via /network/status.
- Get /block response using empty request.
- Compare the response and current tip.

However we call /block after /network/status, so we may have new blocks coming in during the gap, which makes previous tip not the latest block anymore. Therefore instead of saying tip.Index == block.BlockIdentifier.Index, we should provide a buffer (aka tipDiff).

Reply:
Actually I just realized that we have a better way to do this: just calling /block after /network/status. So in this case, we just need to check if the block index returned by /block is >= the index returned by /network/status. This should be more intuitive. I'll make the change.

Reply:
Done.